### ISSUE TITLE
Default values of Adaboost program are not good

### ISSUE THREAD

**Issue by NippunSharma** [mlpack/mlpack#3010]
#### Issue description

This issue was encountered when I was testing the python generated wrapper of Adaboost for my GSoC project for compatibility with scikit hyperparameter tuners in Python.

#### Steps to reproduce

I am pasting the code that I used but the changes that I am working on are still not merged into mlpack.
However, I am also pasting the best parameters and results that I got after tuning. To reproduce this, you can
try to run the currently available `adaboost()` program through python with the given parameters and confirm the results.

```python
from skopt import BayesSearchCV
from mlpack import Adaboost
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score
import numpy as np

X, y = load_digits(n_class=10, return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.75, test_size=.25, random_state=0)

# training a model with default parameters and finding out the score.
modelNotTuned = Adaboost()
modelNotTuned.fit(X_train, y_train)
preds = modelNotTuned.predict(X_test)
f1_score(y_test, preds, average="weighted")
# output:
# 0.05856060206432102 (very bad)

# Now, tuning the model.
opt = BayesSearchCV(
    Adaboost(),
    {
        'iterations': (100, 500),
        'weak_learner': ['perceptron', 'decision_stump']
    },
    n_iter=10,
    cv=3,
    scoring="f1_weighted",
    random_state=1,
)

opt.fit(X_train, y_train)

print("val. score: %s" % opt.best_score_)
print("test score: %s" % opt.score(X_test, y_test))
# output: 
# val. score: 0.9554649920291841
# test score: 0.9558136593062766 (much better)

opt.best_params_
# output:
# OrderedDict([('iterations', 226), ('weak_learner', 'perceptron')])
```
#### Expected behavior
While we do expect hyperparameter tuning to improve the results, default parameters should
at least be able to provide some decent enough results.

#### Actual behavior
The default parameters provided very bad results.

#### Remarks
Some research should be done on finding out a set of parameters that can provide a decent score
in the majority of tasks. For reference, scikit's `AdaBoostClassifier()` gives a weighted f1 score of about
0.19 (on the same dataset shown in the example), so we should aim to achieve at least that much.

**Comment by rcurtin** [comment]
I think, at least for the MNIST example here, part of the reason is that each decision stump can only split two classes (at most) on one feature, so decision stumps will not really work well here.  I would imagine that in the two-class case, even on MNIST-type data, decision stumps would work much better.

Nonetheless, some exploration is worthwhile here: we want to find some good default settings that tend to work well over a variety of datasets.  They don't need to do the best, of course, but we should hopefully manage to do better than the results we are seeing here. :)

**Comment by AditiThirdEye** [comment]
Please assign me this issue if not PR yet.

**Comment by AdhaarSharma** [comment]
Please assign me this issue if not resolved.

**Comment by rcurtin** [comment]
No need to assign it, just work on it and open a PR. :+1: This PR will require a decent bit of investigation about how our defaults perform on different datasets.

**Comment by dnabanita7** [comment]
This issue is outdated. Most of the lines of code have different datatypes and gives error while reproducing this issue.

**Comment by rcurtin** [comment]
@dnabanita7 the code snippets may be outdated, but the underlying issue should still be relevant.  Can you post more information about the problems you are having?

**Comment by dnabanita7** [comment]
@rcurtin I tried to reproduce this [here](https://colab.research.google.com/drive/113UPcZbyh4BMqbF1MkiNAu38unm45Q2H?usp=sharing), but as you can see [BayesSearchCV](https://scikit-optimize.github.io/stable/modules/generated/skopt.BayesSearchCV.html) requires and estimator while in ```d = adaboost(input_model=None, iterations=1000, labels=np.empty([0],
        dtype=np.uint64), test=np.empty([0, 0]), tolerance=1e-10,
        training=np.empty([0, 0]), verbose=False,
        weak_learner='decision_stump')```,  ``d`` is not an estimator. I tried using ``d[output_model]`` instead but it gave soem other errors. 

Apart from that, I tried using one of our [HyperParameterTuner class](https://www.mlpack.org/doc/stable/doxygen/hpt_guide.html) but I couldn't find python binding for this. Also, We only have SimpleCV and KFoldCV implemented.

**Comment by pree-T** [comment]
Is this issue still open? May I work on it? Can someone help me?

**Comment by dnabanita7** [comment]
No need to ask. Open a PR and refer to this issue.

**Comment by rcurtin** [comment]
You don't need to use the hyperparameter tuning or anything.  You can just use the command-line program or the Python bindings to replicate the original finding that convergence is not good for the given dataset.

**Comment by dnabanita7** [comment]
Yes, as per [here](https://colab.research.google.com/drive/113UPcZbyh4BMqbF1MkiNAu38unm45Q2H?usp=sharing), convergence of adaboost is not good for the given dataset.

**Comment by rcurtin** [comment]
@OmkarGowda990 the issue is not resolved, so feel free to work on it.

**Comment by UrmilPawar** [comment]
Hello, I am Urmil Pawar. I am just a beginner in open source contribution but I would like to give a try to your issue.Can you please explain me how to approach it

**Comment by harindra2002** [comment]
<img width="297" alt="Screenshot 2023-02-16 at 3 57 08 PM" src="https://user-images.githubusercontent.com/60562187/219339572-9204f159-a7da-4641-84ac-0c879216a856.png">

I've tested it on multiple datasets, the classifier is doing well when the number of classes is less. Ig that's obvious given that the weak_learner is set to decision stump by default. But otherwise it is working on par with sklearn model

**Comment by harindra2002** [comment]
If weak_learner is set to perceptron then for MNIST the f1 score is 0.95581 and for olivetti_faces it is 0.9118

**Comment by harindra2002** [comment]
`import numpy as np
from mlpack.adaboost import Adaboost
from sklearn.datasets import load_digits,load_iris,load_breast_cancer,load_wine,fetch_olivetti_faces
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score,accuracy_score
from sklearn.ensemble import AdaBoostClassifier

#mnist
print("MNIST 10 classes:")
X, y = load_digits(n_class=10, return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.75, test_size=.25, random_state=0)
modelNotTuned = Adaboost()
modelNotTuned.fit(X_train, y_train)
preds = modelNotTuned.predict(X_test)
print("mlpack")
print("F1: ",f1_score(y_test, preds, average="weighted"))
print("Accuracy: ",accuracy_score(y_test, preds),"\n")


clf=AdaBoostClassifier()
clf.fit(X_train,y_train)
preds=clf.predict(X_test)
print("Sklearn model:")
print("F1: ",f1_score(y_test, preds, average="weighted"))
print("Accuracy: ",accuracy_score(y_test, preds),"\n")


#iris
print("IRIS 3 classes:")
X, y = load_iris(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.75, test_size=.25, random_state=0)
modelNotTuned = Adaboost(weak_learner='perceptron')
modelNotTuned.fit(X_train, y_train)
preds = modelNotTuned.predict(X_test)
print("mlpack")
print("F1: ",f1_score(y_test, preds, average="weighted"))
print("Accuracy: ",accuracy_score(y_test, preds),"\n")


clf=AdaBoostClassifier()
clf.fit(X_train,y_train)
preds=clf.predict(X_test)
print("Sklearn model:")
print("F1: ",f1_score(y_test, preds, average="weighted"))
print("Accuracy: ",accuracy_score(y_test, preds),"\n")

#Breast cancer
print("Breast cancer 2 classes:")
X, y = load_breast_cancer(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.75, test_size=.25, random_state=0)
modelNotTuned = Adaboost(weak_learner='perceptron')
modelNotTuned.fit(X_train, y_train)
preds = modelNotTuned.predict(X_test)
print("mlpack")
print("F1: ",f1_score(y_test, preds, average="weighted"))
print("Accuracy: ",accuracy_score(y_test, preds),"\n")


clf=AdaBoostClassifier()
clf.fit(X_train,y_train)
preds=clf.predict(X_test)
print("Sklearn model:")
print("F1: ",f1_score(y_test, preds, average="weighted"))
print("Accuracy: ",accuracy_score(y_test, preds),"\n")

#Wine
print("Wine 3 classes:")
X, y = load_wine(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.75, test_size=.25, random_state=0)
modelNotTuned = Adaboost(weak_learner='perceptron')
modelNotTuned.fit(X_train, y_train)
preds = modelNotTuned.predict(X_test)
print("mlpack")
print("F1: ",f1_score(y_test, preds, average="weighted"))
print("Accuracy: ",accuracy_score(y_test, preds),"\n")


clf=AdaBoostClassifier()
clf.fit(X_train,y_train)
preds=clf.predict(X_test)
print("Sklearn model:")
print("F1: ",f1_score(y_test, preds, average="weighted"))
print("Accuracy: ",accuracy_score(y_test, preds),"\n")

#fetch_olivetti_faces
print("olivetti_faces 40 classes:")
X,y=fetch_olivetti_faces(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.75, test_size=.25, random_state=0)
modelNotTuned = Adaboost()
modelNotTuned.fit(X_train, y_train)
preds = modelNotTuned.predict(X_test)
print("mlpack")
print("F1: ",f1_score(y_test, preds, average="weighted"))
print("Accuracy: ",accuracy_score(y_test, preds),"\n")

clf=AdaBoostClassifier()
clf.fit(X_train,y_train)
preds=clf.predict(X_test)
print("sklearn model")
print("F1: ",f1_score(y_test, preds, average="weighted"))
print("Accuracy: ",accuracy_score(y_test, preds))
`

**Comment by harindra2002** [comment]
@rcurtin I think this can be closed

**Comment by Anchaliya75** [comment]
@harindra2002 so adaboost will always give poor results if classes are higher if we use decision stump as weak learner?

**Comment by harindra2002** [comment]
yes

**Comment by rcurtin** [comment]
Based on your simulations it looks like the issue still exists.  @NippunSharma's original issue was that we should see if we can change the defaults to perform better across datasets.  In your experiments you showed that the problem exists when there are a greater number of classes; but the question still remains, can we tune the default parameters of the decision stump such that they do a little better?  I feel like 5% accuracy is really insanely bad when compared to the perceptron's result, and even if we are using decision stumps better should be very possible.

**Comment by IWNMWE** [comment]
Hey @rcurtin I have been gong through the issue for a few days now and according to my observations the reason this error is coming up is due to the low performance of the initial stump the alpha value comes down to near zero nearly 7e-15 and even if we set the tolerance low enough it still wont help because due to the the low accuracy decision stump most of the datapoints are miss classified hence causing nearly no change in the updating of D matrix hence it does not improve on consecutive runs by much even recent papers have acknowledged this issue and have come up with better base learners like this paper suggests
https://arxiv.org/pdf/1312.6086.pdf Can you pls explain why the mh algo was preferred over the SAMME as this solves the issue by having a log(k-1) in its alpha equation??

**Comment by IWNMWE** [comment]
Hey @rcurtin @NippunSharma  can we explore making multiple variations of adaboost available like adding samme and samme.r also available to the user cause i have tested by adding samme into the adaboost_impl.hpp and its giving  good results on the digits dataset 
![Screenshot 2023-03-03 175052](https://user-images.githubusercontent.com/118750341/224036998-b3ad43f0-0f27-4e9f-b8e2-b0e98bcdf300.png)

this is the f1 score of the test dataset returned 


These are the files returned:
  
[Testoutput.csv](https://github.com/mlpack/mlpack/files/10932071/Testoutput.csv)

[Uploading digitstargettest.csv…]()


#include <mlpack.hpp>
#pragma warning(disable:4146)
using namespace mlpack;
using namespace std;
using namespace arma;

int main()
{
    
        
    arma::mat datatrain;
    arma::mat datatest;
    arma::Row<size_t> label;
    arma::Row<size_t> labeltest;
    data::Load("C:/Users/shukl/Desktop/digits.csv", datatrain);
    data::Load("C:/Users/shukl/Desktop/digitstarget.csv", label);
    data::Load("C:/Users/shukl/Desktop/digitstest.csv", datatest);
    data::Load("C:/Users/shukl/Desktop/digitstargettest.csv", labeltest);
    const size_t numClasses = 10;
    ID3DecisionStump ds(datatrain, label, numClasses);
    size_t iterations = 100;
    double tolerance = 1e-17;

    AdaBoost<ID3DecisionStump> a(tolerance);
    double ztProduct = a.Train(datatrain, label, numClasses, ds,
        iterations, tolerance);
    arma::Row<size_t> predictedLabels;
    a.Classify(datatest, predictedLabels);
    cout << predictedLabels;
}


This was my main program 

 and this is the adaboost_impl.hpp i used where i only changed the train method

[impl.txt](https://github.com/mlpack/mlpack/files/10932123/impl.txt)


these are the train data files 
[digits.csv](https://github.com/mlpack/mlpack/files/10932128/digits.csv)
[digitstarget.csv](https://github.com/mlpack/mlpack/files/10932130/digitstarget.csv)

this is the test data file

[digitstest.csv](https://github.com/mlpack/mlpack/files/10932134/digitstest.csv)

**Comment by IWNMWE** [comment]
pls remove the cout of D line 166 in the impl file i kept that for checking the algo

**Comment by Jay-sanjay** [comment]
sir , is the issue resolved 
I want to contribute to it

**Comment by prashant-shah01** [comment]
Can anyone suggest me how to start open source contribution  . I am beginner in this field

**Comment by indrasn0wal** [comment]
Is this issue still open?

**Comment by man0045** [comment]
is this issue still open

**Comment by Joyee2004** [comment]
Hello, this is Joyee Biswas. I would like to try working on this issue..  @rcurtin kindly assign me this issue.

**Comment by aamod-wick** [comment]
Hi @rcurtin @NippunSharma i have tried t understand the code base and issues for many days of mlpack but i am unable to do so alone as i am a newbie. Could you please guide me in how i understand the codebase of mlpack or if there is any irc/discord chat /mail?? I apologise if i have broken any etiquette.

**Comment by DreamerChaserHAH** [comment]
@aamod-wick You can check their [community webpage](https://www.mlpack.org/community.html#page-chat) for that
